Basic support for ewasm precompiles

s1na opened the pull request:
This is the first step for integrating ewasm into ethereumjs-vm. It adds very basic support for using [ewasm-precompiles](https://github.com/ewasm/ewasm-precompiles) instead of those written natively in JS. The interface only implements methods which the precompiles use (all of them are sync). The code is heavily inspired by [ewasm-kernel](https://github.com/ewasm/ewasm-kernel) :)

It passes state and blockchain tests (commands `npm run testStateEwasm` and `npm run testBlockchainEwasm`), except for cases which use ripemd160 (that's why I've commented that out for now). The ripemd160 precompile written in rust doesn't pad the output to 32 bytes, and I wasn't sure where whether padding was intentionally left to be done in hosts.

Another point is the wasm binaries. I committed the binaries as well, but I'm not sure if this is the best way to distribute them, and whether using `fs.readFileSync` is a good idea (specially for browser support).

To enable these changes, the option `enableEwasm` needs to be set to `true` when instantiating `VM`.

axic commented:
@s1na wow! Will review shortly.

axic commented on lib/ewasm/system/transform-i64.wast:
```diff
@@ -0,0 +1,13 @@
+(module
```
You could use `wabt` (as an npm module) as part of the build system to generate the `wasm` of this.

s1na replied:
Did you mean this [wabt.js](https://github.com/AssemblyScript/wabt.js)? Seems a bit buggy, haven't been able to get it to work yet. Will keep working on it.

Or rather including the wabt toolkit itself?

axic replied:
Meant wabt.js

axic commented on lib/ewasm/system/transform-i64.wast:
```diff
@@ -0,0 +1,13 @@
+(module
+  (import "interface" "useGas" (func $useGas (param i32 i32)))
+
+  (export "useGas" (func $useGasShim))
```
There's a lot more shims needed. Can you add a comment into this file stating the purpose and that this only implement those currently used, but not all the EEI methods?

axic commented on lib/index.js:
```diff
@@ -40,6 +40,7 @@ VM.deps = {
  * @param {Boolean} opts.activatePrecompiles create entries in the state tree for the precompiled contracts
  * @param {Boolean} opts.allowUnlimitedContractSize allows unlimited contract sizes while debugging. By setting this to `true`, the check for contract size limit of 24KB (see [EIP-170](https://git.io/vxZkK)) is bypassed. (default: `false`; ONLY set to `true` during debugging)
  * @param {Boolean} opts.emitFreeLogs Changes the behavior of the LOG opcode, the gas cost of the opcode becomes zero and calling it using STATICCALL won't throw. (default: `false`; ONLY set to `true` during debugging)
+ * @param {Boolean} opts.enableEwasm enable EWASM precompiles. (default: `false`; Experimental feature)
```
I'd name this `enableEwasmPrecompiles` or something similar, since it is not complete ewasm.

axic commented on lib/ewasm/env.js:
```diff
@@ -0,0 +1,117 @@
+    // convert from a 32-bit two's compliment
+    low = 0x100000000 - low
+  }
+  // JS only bitshift 32bits, so instead of high << 32 we have high * 2 ^ 32
```
Do you have a reference for this?

s1na replied:
I adopted this from [here](https://github.com/ewasm/ewasm-kernel/blob/master/EVMimports.js#L691), and not quite sure if there are no edge cases. But it is working for the test suite. I added the `assert` you wrote in the next comment.

axic commented on lib/ewasm/env.js:
```diff
@@ -0,0 +1,117 @@
+  if (high < 0) {
+    // convert from a 32-bit two's compliment
+    high = 0x100000000 - high
+  }
```
Can you add an assert here that `high` is doesn't have any bits set between 32-21 bits?

(JS can store up to 53 bits)

axic commented on lib/ewasm/contract.js:
```diff
@@ -0,0 +1,32 @@
+      this._instance.exports.main()
+    } catch (e) {
+      if (!(e instanceof VmError) && e.message !== 'unreachable') {
+        throw e
```
Why not rethrow `VmError`?

I think what this should do is only catch a single exception: `FinishExecution`, which is emitted in `finish` (just create a new exception type). Every other exception is real and should not be ignored.

axic replied:
Well you could wrap wasm execution errors into a `VmError` exception if we want to hide wasm exceptions - we may already do that for `opFns` in `runCode`, wrapping them into "internal errors".

s1na replied:
I used `VmError` to remain consistent with `opFns`'s error handling, and it's thrown for out of gas and revert. The reason why I catch this here is because `runCode` doesn't do any error handling, it excepts the error to be in the `results` object. Maybe this is something that can be changed in refactoring?
Is it always the case that ewasm will throw `unreachable` immediately after calling `finish`? In that case yeah it might be better to throw a custom error in `finish`.

axic replied:
There's no requirement to have `unreachable` after `finish` - the finish method should abort execution.

Rust inserts the unreachable opcode after functions which are marked non-returned and `finish` is marked as such in ewasm-rust-api, but this code here should not rely on that.

axic commented on .circleci/config.yml:
```diff
@@ -37,6 +37,15 @@ jobs:
+          at: ~/project
+      - *restore_node_modules
+      - run:
+          name: buildWast
```
This is free text, can have spaces.

axic commented on lib/ewasm/index.js:
```diff
@@ -0,0 +1,25 @@
+
+const precompiles = {
+  '0000000000000000000000000000000000000002': contracts['sha256'],
+  // '0000000000000000000000000000000000000003': contracts['ripemd160'],
```
Need to enable this.

axic commented on lib/ewasm/system/transform-i64.wast:
```diff
@@ -0,0 +1,21 @@
+  ;; Use gas takes a i64 parameter `amount`. This shim takes the i64 value
+  ;; and breaks it into two 32 bit values `high` and `low` which js can natively
+  ;; support.
+  (func $useGasShim
```
Can you format this source code consistently? e.g. 2 or 4 spaces indentation for each level.

s1na commented:
A few small updates:

- `finish` now throws `FinishExecution` which `Contract.run` catches
- Coverage now additionally runs state tests with ewasm precompiles enabled
- Fixed indentation of `transform-i64.wast`
- Added some jsdoc comments to `Contract` and `Env`

P.S. Sorry that I keep rebasing. If that makes reviewing more difficult I can keep full commit history, and rebase once at the end.

axic commented:
I think this looks great, but I am not at all sure merging in those big binaries. They supposed to be way smaller once we fix an optimisation step. Ideally compilation of would happen at build time, but unfortunately that would put a requirement on rust available.

Perhaps the best way is to release the binaries on the ewasm-precompiles repo and the build script here would download them with curl.

@s1na what do you think?

s1na commented:
If I'm not mistaken, github allows distributing files for release (if having them in git is not desirable). It definitely wouldn't hurt to have the binaries downloadable from ewasm-precompiles.

I'd say if the precompiles change very frequently, fetching them via curl in the build scripts is much better, otherwise there are advantages and disadvantages to both approaches.

If the optimization is close, we could also wait and check in the smaller binaries.

axic commented:
> If I'm not mistaken, github allows distributing files for release (if having them in git is not desirable).

Yes, to clarify that is what I meant. Tagging the repo and making a proper release on github - there one can upload files.

s1na commented:
> Tagging the repo and making a proper release on github

Is there anything I can help with for that?

I was wondering if publishing them as an npm package would be better in any way? Or alternatively, could the rust compiler produce `wast` files in addition to wasm binaries?

axic commented:
> Is there anything I can help with for that?

I was waiting for the modexp precompile to be merged to make a "stable" release.

holgerd77 commented:
Do I get this correctly that this didn't stand the test of time and we therefore might want to close here? 🤔 Or are there remaining applications and/or use cases that I am missing?
